### ISSUE TITLE
Having to manually recast output or target (e.g. "expected Tensor but got HalfTensor")

### ISSUE THREAD

**Issue by drscotthawley** [NVIDIA/apex#195]
> "With Amp, it isn't necessary to manually convert data to half."
(^from line 263, https://github.com/NVIDIA/apex/blob/master/examples/imagenet/main_amp.py)

> "Users should not manually cast their model or data to .half(), regardless of what opt_level or properties are chosen."
(^ from https://nvidia.github.io/apex/amp.html)

I'm getting and error if I leave the targets uncast. Here's a minimal 'schematic' representation of what my code is doing, based on calculating log-cosh error:

```
model = MyModel()
model.to(device)
optimizer = torch.optim.Adam(list(model.parameters()), lr=lr_sched[0], weight_decay=0)
model, optimizer = amp.initialize(model, optimizer, opt_level="O3")

for x, y_target in my_dataloader:
    x, y_target = x.to(device), y_target.to(device)
    y_pred = model.forward(x)
    loss = torch.mean( torch.log( torch.cosh(y_target - y_pred) ))
    optimizer.zero_grad()
    with amp.scale_loss(loss, optimizer) as scaled_loss:
         scaled_loss.backward()
    torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), max_norm=1., norm_type=1)
    model.clip_grad_norm_()   
    optimizer.step()
```
I'm finding that I get a Runtime error,...

`RuntimeError: expected type torch.cuda.FloatTensor but got torch.cuda.HalfTensor
`
unless I manually convert the "y_target" to half, as in...

 `loss = torch.mean( torch.log( torch.cosh(y_target.half() - y_pred) ))`

...this runs but then I get NaNs.   Alternatively, manually casting y_pred as `float` in the loss calculation also runs but also produces NaNs.

The need to manually recast targets seems to be in contradiction to the docs and ImageNet example.  

Can someone clarify this?  Thanks.    

(I see other Issues posted related to this but they seem to involve FusedAdam or BatchNorm or FP16_Optimizer, none of which I'm using.)  

PS- The only thing that runs and produces ordinary numbers for me so far, is to revert back to the non-amp-related code.

**Comment by drscotthawley** [comment]
Seems the NaNs are due to opt_level="O3".  Changing to "O1" or "O2" produces a seemingly unending stream of "gradient overflow" errors, whereby the loss scale is reduced to 1e-200 and below.   (?)

**Comment by mcarilli** [comment]
In general, it's always good practice to manually call .float() on anything coming into a loss function.  This will be a no-op if the data is already float, and the loss computation is usually a negligible part of your network runtime, so there's no reason not to call .float() on arguments to the loss.  That being said, your original point is still valid:  I shouldn't require you to make this cast manually.  I'll change the model-forward patching so that the output is cast to float with "O2" and "O3" (with "O1," this casting should be taken care of by the loss function itself).

Gradient overflows are not a separate issue from the the NaNs/infs.  Reducing the loss scale is how Amp (with dynamic loss scaling) reacts when it encounters NaNs, so you're seeing an alternative symptom of the same disease.

When clipping gradients, you should not have to clip the model grads directly.  Clipping master grads is done _instead of_ (as opposed to _in addition to_) clipping the model grads.  If the master params are completely distinct from the model params, then clipping the model grads doesn't do any harm (it doesn't affect anything, because the model grad->master grad copy has already occurred on context manager exit, and only the master grads are used by the optimizer).   However, some of the model params may also be master params.  Specifically, any model params that are FP32 are used directly as master params (in the case of "O2," this is the batchnorm params, in the case of "O1," all the params are left in FP32, so the things called "model params" and "master params" fully coincide).  If you manually clip both the "model params" and the "master params", any params that coincide in the two lists will be clipped twice.

It looks like you were able to find [my docs on gradient clipping](https://nvidia.github.io/apex/advanced.html#gradient-clipping), which is good, but on rereading them they are not clear on this point, which may be the source of your nans.  Try removing the `model.clip_grad_norm_()` line (i'm not sure where that's coming from, i don't see that method defined on `torch.nn.module` by default, but in any case it should not be needed) and rerun with `opt_level="O1"`.

**Comment by mcarilli** [comment]
[Updated documentation](https://nvidia.github.io/apex/advanced.html#gradient-clipping), still not 100% sure this is your issue though.

**Comment by drscotthawley** [comment]
Wow, thanks so much @mcarilli for the detailed response!   

I get the unending "Gradient overflow" messages whether I call my model's `clip_grad_norm_()` method or not, running with opt_level="O1". 

>  i don't see that method defined on torch.nn.module by default

Yea, not by default.  My model actually does have its own clip_grad_norm_() that my collaborator and I defined, which refers to parameters that are part of a sub-module called `dft_analysis`:

```
    def clip_grad_norm_(self):
        torch.nn.utils.clip_grad_norm_(list(self.dft_analysis.parameters()) +
                                      list(self.dft_synthesis.parameters()),
                                      max_norm=1., norm_type=1)
```

But my optimizer should include these, I'd think, because it's defined in terms of model.parameters()...

`optimizer = torch.optim.Adam(list(model.parameters()), lr=lr_sched[0], weight_decay=0)`



Here's the full picture of what I've got right now:


```
            # Optimization
            optimizer.zero_grad()

            #loss.backward()   # replace this with amp call 
            with amp.scale_loss(loss, optimizer) as scaled_loss:
                scaled_loss.backward()
            torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), max_norm=1., norm_type=1)  # this makes it work

            # Clip model norm
            ''' commenting out for now
            if parallel:
                for child in model.children():  # for DataParallel
                    if hasattr(child, 'clip_grad_norm_'):
                        child.clip_grad_norm_()
            else:
                if hasattr(model, 'clip_grad_norm_'):
                    model.clip_grad_norm_()   # for non-DataParallel
            '''

            optimizer.step()
```




Here's a more detailed log of what I'm getting right now:

```
Selected optimization level O1:  Insert automatic casts around Pytorch functions and Tensor methods.

Defaults for this optimization level are:
enabled                : True
opt_level              : O1
cast_model_type        : None
patch_torch_functions  : True
keep_batchnorm_fp32    : None
master_weights         : None
loss_scale             : dynamic
Processing user overrides (additional kwargs that are not None)...
After processing overrides, optimization options are:
enabled                : True
opt_level              : O1
cast_model_type        : None
patch_torch_functions  : True
keep_batchnorm_fp32    : None
master_weights         : None
loss_scale             : dynamic

Gradient overflow.  Skipping step, reducing loss scale to 32768.0
Gradient overflow.  Skipping step, reducing loss scale to 16384.0
Gradient overflow.  Skipping step, reducing loss scale to 8192.0
Gradient overflow.  Skipping step, reducing loss scale to 4096.0
Gradient overflow.  Skipping step, reducing loss scale to 2048.0
Gradient overflow.  Skipping step, reducing loss scale to 1024.0
Gradient overflow.  Skipping step, reducing loss scale to 512.0
Gradient overflow.  Skipping step, reducing loss scale to 256.0
Gradient overflow.  Skipping step, reducing loss scale to 128.0
epoch 1/1000, time: 2.90: lr = 6.67e-06, data_point 2000: loss: 5.370e-03   Gradient overflow.  Skipping step, reducing loss scale to 64.0
Gradient overflow.  Skipping step, reducing loss scale to 32.0
Gradient overflow.  Skipping step, reducing loss scale to 16.0
Gradient overflow.  Skipping step, reducing loss scale to 8.0
Gradient overflow.  Skipping step, reducing loss scale to 4.0
Gradient overflow.  Skipping step, reducing loss scale to 2.0
Gradient overflow.  Skipping step, reducing loss scale to 1.0
Gradient overflow.  Skipping step, reducing loss scale to 0.5
Gradient overflow.  Skipping step, reducing loss scale to 0.25
Gradient overflow.  Skipping step, reducing loss scale to 0.125
epoch 1/1000, time: 3.52: lr = 6.67e-06, data_point 4000: loss: 5.251e-03   Gradient overflow.  Skipping step, reducing loss scale to 0.0625
Gradient overflow.  Skipping step, reducing loss scale to 0.03125
Gradient overflow.  Skipping step, reducing loss scale to 0.015625
Gradient overflow.  Skipping step, reducing loss scale to 0.0078125
Gradient overflow.  Skipping step, reducing loss scale to 0.00390625
Gradient overflow.  Skipping step, reducing loss scale to 0.001953125
Gradient overflow.  Skipping step, reducing loss scale to 0.0009765625
Gradient overflow.  Skipping step, reducing loss scale to 0.00048828125
Gradient overflow.  Skipping step, reducing loss scale to 0.000244140625
...and these messages keep going and the scale gets arbtrarily small. 
```

Also, changing out the log-cosh loss function for MAE or MSE has no effect; same messages. 


If you have any other thoughts, I'm all ears. Obviously you can't support every user's issues, but I'll pass on whatever I learn to others! 
Any other ideas?

**Comment by drscotthawley** [comment]
Found it. The one line in my model that makes all the difference:

`phase = torch.atan2(x_imag, x_real+1e-6)
`
I need to "float" both x_imag and x_real explicitly.  Yes, the 1e-6 was causing catastrophic loss of precision and x_real.float() prevents NaNs, but the x_imag also needs float-ing or else the error is way too high. 

If I do this, the loss scale stops at 8.0.  And either "O1" or "O2" optimizations work fine. 

Two more questions, if I may:

1.  Is torch.atan2 on the list of whitelisted functions?  If not, maybe it should be. 

2. Not seeing much speedup.   I did build Apex with the C++ and CUDA flags enabled (CUDA 10 is installed), but  I had just grabbed the "conda" version of PyTorch for CUDA 10.  Do I also need to build PyTorch from source in order to see a significant speedup for FP16?     (Details: "O0" runs in 88 seconds per epoch for my app, "O1" runs in 81 seconds and "O2" is 78 seconds.  This is on an RTX2080 Ti).

**Comment by drscotthawley** [comment]
This is worth it's own comment: Watched Michael's webinar, and noted the "Tips" included (on slide page 29) using `torch.backends.cudnn.benchmark = True`. 
*This* gave me a factor of 2 speedup overall -- which is awesome! -- however both "O0" and "O2" still yield essentially the same execution time.

**Comment by mcarilli** [comment]
It appears atan2 is not on the FP32-ified list (which we call the blacklist), and apparently it should be.  We add things to the FP32-ified list based on experience, so this is extremely valuable feedback.

It sounds like everything you're doing is reasonable.  CPU-side profiling is not going to tell you much because GPU execution is (hopefully) mostly asynchronous.  The CPU-side timings will tell you how long it took to launch the kernels, but not how long they actually took to run.  Imo visual profiling is a good path to truth:  https://gist.github.com/mcarilli/213a4e698e4a0ae2234ddee56f4f3f95.  This is for distributed profiling, but the tips it gives about "focusing" your profiling with cudaProfilerStart/Stop are still relevant.  For single-process profiling you can just do
```
nvprof --profile-from-start off -fo output.nvprof python myscript.py args....
```
then open Nvidia Visual Profiler and import `output.nvprof`.  Look for big kernels or big whitespace gaps on the device timeline.

**Comment by GadirajuSanjayvarma** [comment]
@mcarilli @drscotthawley 
Hello sir.I am training on a model which will predict depth images and mask images.
My training code:
```
class Train:
  def __init__(self, model, dataloader, optimizer, stats, scheduler=None, L1lambda = 0,criterion=None,use_amp=True):
    self.model = model
    self.dataloader = dataloader
    self.optimizer = optimizer
    self.scheduler = scheduler
    self.stats = stats
    self.L1lambda = L1lambda
    self.criterion=criterion
    self.loss1=0.0
    self.loss2=0.0
    self.loss=0.0
    self.use_amp=True
  def run(self):
    self.model.train()
    torch.cuda.empty_cache()
    pbar = tqdm_notebook(self.dataloader)
    for data1,data2,target1,target2 in pbar:
      # get samples
      data1,data2 = data1.to(self.model.device), data2.to(self.model.device)
      target1, target2 = target2.to(self.model.device), target2.to(self.model.device)
      self.optimizer.zero_grad()
      output1,output2 = self.model(data1,data2)
      #print(target.shape)
      target1=target1.unsqueeze(1)
      target2=target2.unsqueeze(1)
      self.loss1=self.criterion[0](output1, target1)
      self.loss2=self.criterion[1](output2, target2)
      #print("loss1 {}".format(self.loss1))
      #print("loss2 {}".format(self.loss2))
      self.loss=(self.loss1+self.loss2)/2.0
      
      # In PyTorch, we need to set the gradients to zero before starting to do backpropragation because PyTorch accumulates the gradients on subsequent backward passes. 
      # Because of this, when you start your training loop, ideally you should zero out the gradients so that you do the parameter update correctly.

      # Predict
      
      #Implementing L1 regularization
      #print(self.loss)
      if self.L1lambda > 0:
        reg_loss = 0.
        for param in self.model.parameters():
          reg_loss += torch.sum(param.abs())
        self.loss += self.L1lambda * reg_loss

      #print(self.loss)
      # Backpropagation
      if self.use_amp:
        with amp.scale_loss(self.loss, self.optimizer) as scaled_loss:
          scaled_loss.backward()
      else:
        self.loss.backward()
      #self.loss.backward()
      #self.optimizer.step()

      # Update pbar-tqdm
      #pred = y_pred.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
      '''n_digits=1
      output = (output * 10**n_digits).round() / (10**n_digits)
      target = (target * 10**n_digits).round() / (10**n_digits)
      correct = output.eq(target).sum().item()
      print("correct is {}".format(correct))'''
      correct=0
      lr = 0
      if self.scheduler:
        lr = self.scheduler.get_last_lr()[0]
      else:
        # not recalling why i used sekf.optimizer.lr_scheduler.get_last_lr[0]
        lr = self.optimizer.param_groups[0]['lr']
      
      #lr =  if self.scheduler else (self.optimizer.lr_scheduler.get_last_lr()[0] if self.optimizer.lr_scheduler else self.optimizer.param_groups[0]['lr'])
      
      self.stats.add_batch_train_stats(self.loss.item(),correct, len(data1), lr)
      pbar.set_description(self.stats.get_latest_batch_desc())
      if self.scheduler:
        self.scheduler.step()

```
My train function calling code:
```
import torch.optim as optim
import torch.nn as nn
import time
EPOCHS =1
#max_lr=1
optimizer = optim.SGD(model.parameters(), lr=0.01)
model, optimizer = amp.initialize(
   model, optimizer, opt_level="O2", 
   keep_batchnorm_fp32=True, loss_scale="dynamic"
)
#criterion=nn.L1Loss()
criterion=[nn.BCEWithLogitsLoss(),nn.L1Loss()]
#criterion = nn.BCEWithLogitsLoss()
start=time.time()
model.gotrain(optimizer, train_loader, test_loader, EPOCHS, "/content/gdrive/My Drive",criterion,use_amp=True)
print(time.time()-start)
```
I am getting an runtime error like this:


RuntimeError: Found param conv1.0.weight with type torch.cuda.HalfTensor, expected torch.cuda.FloatTensor.
When using amp.initialize, you do not need to call .half() on your model
before passing it, no matter what optimization level you choose.

Thank you reading.

**Comment by GadirajuSanjayvarma** [comment]
@drscotthawley 
Okay i rectified it but the training speeds are not improved at all.You said that you added a line
```torch.backends.cudnn.benchmark = True.```.Can you tell me where i should use it in my code for increase in training speed
